exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java

```diff
@@ -69,6 +70,9 @@
 public class HashJoinBatch extends AbstractBinaryRecordBatch<HashJoinPOP> {
   protected static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(HashJoinBatch.class);
 
+  private static final int RECORDS_PER_BATCH = 1024; // internal batches
```

Comment:
These batch sizes are a heuristic. For some data sets these batch sizes will be too small (for example a batch containing nothing but an int column). For other data sets these batch sizes will be too large (batches with 200 int columns). If we really want to accurately manage memory we need to use Paul's batch sizing framework so that we can have a well defined limit on the amount of memory each batch consumes.

Reply:
This is temporary, just to get the spill to happen easily. Also later there's temporary code that just uses a small number of batches (4) to trigger a spill. This is all temporary; after we get the spill mechanism to work, we'd have to start adjusting and tuning all that (that may be the tricky part).

Reply:
Fair enough. Ideally we won't do any tuning though. If we know how much memory we are consuming and limit our batch sizes we can make precise decisions. Paul separated out the memory estimations for the Sort operator into a calculator class which could be easily unit tested and behaved deterministically. Perhaps we can use that in conjunction with the batch sizing framework.